pkg/sqlcache/informer/listoption_indexer.go

```diff
@@ -204,10 +249,94 @@ func NewListOptionIndexer(ctx context.Context, fields [][]string, s Store, names
 }
 
 func (l *ListOptionIndexer) Watch(ctx context.Context, opts WatchOptions, eventsCh chan<- watch.Event) error {
-	key := l.addWatcher(eventsCh, opts.Filter)
+	l.latestRVLock.RLock()
+	latestRV := l.latestRV
+	l.latestRVLock.RUnlock()
+
+	targetRV := opts.ResourceVersion
+	if opts.ResourceVersion == "" {
+		targetRV = latestRV
+	}
+
+	var events []watch.Event
+	var key *watchKey
+	// Even though we're not writing in this transaction, we prevent other write to SQL
+	// because we don't want to add more events while we're backfilling events, so we don't miss events
+	err := l.WithTransaction(ctx, true, func(tx transaction.Client) error {
+		rowIDRows, err := tx.Stmt(l.findEventsRowByRVStmt).QueryContext(ctx, targetRV)
+		if err != nil {
+			return &db.QueryError{QueryString: l.listEventsQuery, Err: err}
+		}
+		if !rowIDRows.Next() && targetRV != latestRV {
+			return fmt.Errorf("resourceversion too old")
+		}
+
+		var rowID int
+		rowIDRows.Scan(&rowID)
+
+		// Backfilling previous events from resourceVersion
+		rows, err := tx.Stmt(l.listEventsStmt).QueryContext(ctx, rowID)
+		if err != nil {
+			return &db.QueryError{QueryString: l.listEventsQuery, Err: err}
+		}
+
+		for rows.Next() {
+			var typ, rv string
+			var buf sql.RawBytes
+			err := rows.Scan(&typ, &rv, &buf)
+			if err != nil {
+				return fmt.Errorf("scanning event row: %w", err)
+			}
+
+			example := &unstructured.Unstructured{}
+			val, err := fromBytes(buf, reflect.TypeOf(example))
+			if err != nil {
+				return fmt.Errorf("decoding event object: %w", err)
+			}
+
+			obj, ok := val.Elem().Interface().(runtime.Object)
+			if !ok {
+				continue
+			}
+
+			if !matchFilter(opts.Filter.ID, opts.Filter.Namespace, opts.Filter.Selector, obj) {
```

Comment:
Probably can hoist `filter := opts.Filter` before the `for rows.Next()` stmt and change this to

`if !matchFilter(filter.ID, filter.Namespace, filter.Selector, obj) {`

The compiler is probably doing this anyway -- this is for human readability, so we don't have to repeatedly parse `opts.Filter`